### PR TITLE
Provide host load indication and bug fixes

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
Host measures load once in a few seconds and caches the result. Clients update meters periodically. This can be made configurable.

#2707 Publish With Options - to Project not working
- should receive list of files rather that erroneously looking into target project

#2708 Connection manager (workspaces) doesn't respect VS theme
- themed control template

#2571 Colorization gets broken in REPL 
- concurrency issues caused replacement at the wrong position.  BufferChanged must come right after replacement character is inserted so we always replace at the end of the buffer.

#2186 IntelliSense support for data.table objects

### CODE REVIEW

File: src/R/Components/Impl/InteractiveWorkflow/Implementation/RInteractiveEvaluator.cs [hunk 1]
```diff
@@ -248,49 +237,59 @@ public sealed class RInteractiveEvaluator : IInteractiveEvaluator {
...
+                // Note: DispatchOnUIThread is expensive, and can saturate the message pump when there's a lot of output,
+                // making UI non-responsive. So avoid using it unless we need it - and we only need it for FlushOutput,
+                // and we only need it to handle CR.
+                if (message.Length > 1 && message[0] == '\r' && message[1] != '\n') {
+                    _coreShell.DispatchOnUIThread(() => {
+                        CurrentWindow.FlushOutput();
+                        // If message starts with CR we remember current output buffer
+                        // length so we can continue writing lines into the same spot.
+                        // See txtProgressBar in R.
+                        if (message.Length > 1 && message[0] == '\r' && message[1] != '\n') {
```
Comment (AlexanderSher) [inline src/R/Components/Impl/InteractiveWorkflow/Implementation/RInteractiveEvaluator.cs]: Looks like a merge conflict - this condition is checked twice

File: src/R/Components/Impl/InteractiveWorkflow/Implementation/RInteractiveEvaluator.cs [hunk 2]
```diff
@@ -248,49 +237,59 @@ public sealed class RInteractiveEvaluator : IInteractiveEvaluator {
...
         /// 'carriage return' i.e. writing into the same line
         /// </summary>
-        private volatile string _lastMessage;
-        private int _lastPosition = -1;
+        struct MessagePos {
+            public string Message;
+            public int Position;
+        }
 
+        private readonly Stack<MessagePos> _messageStack = new Stack<MessagePos>();
```
Comment (AlexanderSher) [inline src/R/Components/Impl/InteractiveWorkflow/Implementation/RInteractiveEvaluator.cs]: Why it is a stack and not a queue?

File: src/R/Components/Impl/Information/HostLoadIndicatorViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        }
+
+        private void OnTimer(object sender, ElapsedEventArgs e) {
+            if (!_disposed && !_busy) {
+                UpdateModelAsync().DoNotWait();
+            }
+        }
+
+        internal Task UpdateModelAsync() {
+            _busy = true;
```
Comment (AlexanderSher) [inline src/R/Components/Impl/Information/HostLoadIndicatorViewModel.cs]: What is this flag for?

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs [hunk 4]
```diff
@@ -66,9 +72,13 @@ internal class ConnectionManager : IConnectionManager {
...
         private async Task CompleteInitializationAsync() {
             await _shell.SwitchToMainThreadAsync();
-            _disposableBag.Add(_statusBar.AddItem(new ConnectionStatusBar {
-                DataContext = _statusBarViewModel
-            }));
+            _disposableBag
+                .Add(_statusBar.AddItem(new ConnectionStatusBar {
+                    DataContext = _statusBarViewModel
+                }))
+                .Add(_statusBar.AddItem(new HostLoadIndicator() {
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: Parentheses aren't required

File: src/Host/Broker/Impl/About/AboutController.cs [hunk 5]
```diff
@@ -14,7 +14,7 @@
 
 namespace Microsoft.R.Host.Broker.About {
     [Authorize(Policy = Policies.RUser)]
-    [Route("/about")]
```
Comment (int19h) [inline src/Host/Broker/Impl/About/AboutController.cs]: Why this change? We've used lowercase for HTTP paths everywhere else, and it is also the standard REST convention.

Comment (int19h) [inline src/Host/Broker/Impl/About/AboutController.cs]: If you just want to distinguish different things, add subpoints under `/about`  (and maybe rename it to `/info`). So we can have `/info/host`, `/info/load` etc.

Although I'd make it `/status/load` etc instead. `/about` or `/info` should be for static stuff that doesn't change for a particular instance of broker, like version numbers or install locations.

File: src/R/Components/Impl/InteractiveWorkflow/Implementation/RInteractiveEvaluator.cs (hunk 2 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/InteractiveWorkflow/Implementation/RInteractiveEvaluator.cs]: So in case we do get multiple changes they will be  applied in reverse order so positions won't change for change 1 after change 2 is applied.

File: src/R/Components/Impl/Information/HostLoadIndicatorViewModel.cs (hunk 3 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/Information/HostLoadIndicatorViewModel.cs]: So we ignore timer events while `GetHostInformationAsync` is sill being executed.

File: src/R/Components/Impl/InteractiveWorkflow/Implementation/RInteractiveEvaluator.cs (hunk 1 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/InteractiveWorkflow/Implementation/RInteractiveEvaluator.cs]: Yep. fixed.

**Comment by MikhailArkhipov** [comment]
Adding picture

image

File: src/R/Components/Impl/InteractiveWorkflow/Commands/SessionInformationCommand.cs [hunk 6]
```diff
@@ -40,27 +41,29 @@ public sealed class SessionInformationCommand : IAsyncCommand {
         }
 
         private async Task PrintBrokerInformationAsync(bool reportTelemetry = false) {
-            var a = await _interactiveWorkflow.RSessions.Broker.GetHostInformationAsync();
-            var window = _interactiveWorkflow.ActiveWindow?.InteractiveWindow;
+            var a = await _interactiveWorkflow.RSessions.Broker.GetHostInformationAsync<AboutHost>();
+            if (a != null) {
```
Comment (MikhailArkhipov) [inline src/R/Components/Impl/InteractiveWorkflow/Commands/SessionInformationCommand.cs]: OK

File: src/R/Components/Impl/Information/HostLoadIndicatorViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            return _interactiveWorkflow.RSessions.Broker.GetHostInformationAsync<HostLoad>()
+                .ContinueWith((t) => {
```
Comment (MikhailArkhipov) [inline src/R/Components/Impl/Information/HostLoadIndicatorViewModel.cs]: OK

File: src/R/Components/Impl/InteractiveWorkflow/Implementation/RInteractiveEvaluator.cs (hunk 2 above)
Comment (AlexanderSher) [inline src/R/Components/Impl/InteractiveWorkflow/Implementation/RInteractiveEvaluator.cs]: Does it mean that only one change may apply?

File: src/R/Components/Impl/Information/HostLoadIndicatorViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        private void OnTimer(object sender, ElapsedEventArgs e) {
+            if (!_disposed && !_busy) {
+                UpdateModelAsync().DoNotWait();
+            }
+        }
+
+        internal async Task UpdateModelAsync() {
+            _busy = true;
+            try {
+                var result = await _interactiveWorkflow.RSessions.Broker.GetHostInformationAsync<HostLoad>();
```
Comment (AlexanderSher) [inline src/R/Components/Impl/Information/HostLoadIndicatorViewModel.cs]: Since we have this code that pings broker all the time, I'd rather move it into `RSessionProvider` and raise an event every time it receives a message. It will allow us to get rid of the expensive `RSessionProvider.RSessionOnDisconnectedAsync`

And you don't need a timer. `TaskScheduler` will do it for you:
```{csharp}
while (_isDisposed) {
    await Task.Delay(interval, cancellationToken);
    var info = await Broker.GetHostInformationAsync<HostLoad>(cancellationToken);
}
```

**Comment by int19h** [comment]
👍

File: src/R/Components/Impl/Information/HostLoadIndicatorViewModel.cs (hunk 8 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/Information/HostLoadIndicatorViewModel.cs]: OK. It is a bit more work, so I am going to open separate item. I guess you want some event like HostStateChanged (live/dead) fired when load cannot be retrieved and we'd use it instead of testing connection. #2722
